pdns/dnsdist.cc

```diff
@@ -1378,13 +1378,14 @@ static void processUDPQuery(ClientState& cs, LocalHolders& holders, const struct
 #endif /* defined(HAVE_RECVMMSG) && defined(HAVE_SENDMMSG) && defined(MSG_WAITFORONE) */
       /* we use dest, always, because we don't want to use the listening address to send a response since it could be 0.0.0.0 */
       sendUDPResponse(cs.udpFD, reinterpret_cast<char*>(dq.dh), dq.len, dq.delayMsec, dest, *dq.remote);
+      memcpy(query, queryBackup, len);
```

Comment:
As stated before queryBackup is a string copy of the original query. We need this copy in case we need to send a reply to user and a request to backend from the same original message. This because the main checking function changes data corresponding to the destination(user/backend). But if the message isn't destined to be sent back to the user, there will be no changes to the querys data. therefor we don't need to reset the querys data since it didn't change. The parent function of this memcpy is the responsible to send message back to user, so if we didn't send anything back the user we dont need to reset it.

Reply:
What modifications are done to the original query and where?
Did you confirm that reverting the original buffer doesn't have an undesired side effect elsewhere?

Reply:
The modification made to the buffer were inside ProcessQuery(now broken into sendReplyCheck and sendBackendCheck). So what i did is basically let the function do the first half of ProcessQuery(sendReplyCheck) and send the message if needed. Then revert it the the start(with queryBackup) and run the second half of ProcessQuery(sendBackendCheck) and send the message if needed.